Editorial review: Add docs for scroll snap events (#36057)
mdn/content: files/en-us/web/api/snapevent/index.md

{{InheritanceDiagram}}

## Constructor

- {{domxref("SnapEvent.SnapEvent", "SnapEvent()")}} {{Experimental_Inline}}
  - : Creates a new `SnapEvent` object instance.

## Instance properties

_Inherits properties from its parent, {{DOMxRef("Event")}}._

- {{domxref("SnapEvent.snapTargetBlock", "snapTargetBlock")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : Returns a reference to the element snapped to in the block direction when the event fired.
- {{domxref("SnapEvent.snapTargetInline", "snapTargetInline")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : Returns a reference to the element snapped to in the inline direction when the event fired.

## Examples

### `scrollsnapchanging` example

In the following `scrollsnapchanging` handler function snippet, we set the {{domxref("SnapEvent.snapTargetBlock", "snapTargetBlock")}} element's `class` attribute to `pending` using the {{domxref("Element.className")}} property, which could be used to style the element differently when it becomes a pending snap target.

> Note that this handler is intended to be set on a block-direction scroll container (vertically-scrolling if the page is set to a left-to-right {{cssxref("writing-mode")}}), therefore only the `snapTargetBlock` element will change between multiple events firing. The element snapped to in the inline direction will always be the same, therefore {{domxref("SnapEvent.snapTargetInline")}} will always contain a reference to the same element.

Author: updated

```js
scrollingElem.addEventListener("scrollsnapchanging", (event) => {
  // Set current pending snap target class to "pending"
  event.snapTargetBlock.className = "pending";

  // Logs the id of the new block-direction snap target element
  console.log(event.snapTargetBlock.id);

  // Always logs the same ID; the inline-direction snap target

>   // element will always be the same

Author: So there was a bit of confusion here. Originally when I started playing with these events, both the `snapTargetBlock` and `snapTargetInline` properties always returned an element reference, even if no snapping was occurring in one direction.

However, I've just tested it again, and Chrome is now returning `null` for the property representing the direction in which no snapping is occurring. Which makes a lot more sense.

This is also basically what the spec says — I must have missed it before, because it doesn't explicitly mention this behavior. However, the properties are marked as nullable, which is what we'd expect.

I've rewritten this bit to suit, and I've also gone through the entire PR to make sure that this behavior is represented correctly in all appropriate places.

  console.log(event.snapTargetInline.id);
});
```

### `scrollsnapchange` example

In the following `scrollsnapchange` handler function snippet, we set a `selected` class on the {{domxref("SnapEvent.snapTargetBlock")}} element, which could be used to style a newly-selected snap target to look like it has been selected (for example, with an animation).

```js
scrollingElem.addEventListener("scrollsnapchange", (event) => {
  event.snapTargetBlock.className = "selected";
});
```

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - {{domxref("Element/scrollsnapchanging_event", "scrollsnapchanging")}} and {{domxref("Element/scrollsnapchange_event", "scrollsnapchange")}} events

Author: I'll keep em.

---
title: SnapEvent
slug: Web/API/SnapEvent
page-type: web-api-interface
status:
  - experimental
browser-compat: api.SnapEvent
---

{{APIRef("Snap Events")}}{{SeeCompatTable}}

> The **`SnapEvent`** interface defines the event object for the {{domxref("Element/scrollsnapchanging_event", "scrollsnapchanging")}} and {{domxref("Element/scrollsnapchange_event", "scrollsnapchange")}} events, which are related to the [CSS scroll snap module](/en-US/docs/Web/CSS/CSS_scroll_snap). Respectively, these fire when the browser determines that a new snap target is pending (will be selected when the current scroll gesture ends), and when a new snap target is selected.
> 
> These can be used to run code in response to new elements being snapped to; `SnapEvent` exposes references to the element snapped to in the inline and/or block direction.

Reviewer: ```suggestion
The **`SnapEvent`** interface defines the event object for the {{domxref("Element/scrollsnapchanging_event", "scrollsnapchanging")}} and {{domxref("Element/scrollsnapchange_event", "scrollsnapchange")}} events. Respectively, these fire when the browser determines that a new snap target is pending (will be selected when the current scroll gesture ends), and when a new snap target is selected.

Defined in the [CSS scroll snap module](/en-US/docs/Web/CSS/CSS_scroll_snap), these can be used to run code in response to new elements being snapped to; `SnapEvent` exposes references to the element snapped to in the inline and/or block direction.
```
move the module out of all the first sentences. where it's defined doesn't need to show up in search, devtools, etc.

Author: updated

_Inherits properties from its parent, {{DOMxRef("Event")}}._

- {{domxref("SnapEvent.snapTargetBlock", "snapTargetBlock")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : Returns a reference to the element snapped to in the block direction when the event fired, or `null` if no element is snapped to in the block direction.
- {{domxref("SnapEvent.snapTargetInline", "snapTargetInline")}} {{ReadOnlyInline}} {{Experimental_Inline}}

>   - : Returns a reference to the element snapped to in the inline direction when the event fired, or `null` if no element is snapped to in the inline direction.

Reviewer: in my tests, i get a reference to the first element. I haven't been able to get a null

Reviewer: OK, was able to get a null, but only if there was no way to snap in the inline direction. If there is the possibility of inline scrolling as well as block scrolling, i get a value for both. If you can only scroll in one direction, the other direction is null. Should we clarify?

Author: It's a really good idea to call this out explicitly. I've added an explanation of this on the `SnapEvent` page and on the scroll snap events guide page. I've also mentioned it on the `snapTargetBlock`/ `snapTargetInline` pages.
